test/e2e/layering_test.go

```diff
@@ -0,0 +1,98 @@
+package e2e_test
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	buildv1 "github.com/openshift/api/build/v1"
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/openshift/machine-config-operator/test/constants"
+	"github.com/openshift/machine-config-operator/test/framework"
+	"github.com/openshift/machine-config-operator/test/helpers"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestBootInClusterImage(t *testing.T) {
+	cs := framework.NewClientSet("")
+
+	// create a new image stream
+	imageStreamName := "test-boot-in-cluster-image"
+	buildName := imageStreamName
+	imageStreamConfig := &imagev1.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      imageStreamName,
+			Namespace: constants.MCONamespace,
+		},
+	}
+	_, err := cs.ImageStreams(constants.MCONamespace).Create(context.TODO(), imageStreamConfig, metav1.CreateOptions{})
+	require.Nil(t, err)
+	defer cs.ImageStreams(constants.MCONamespace).Delete(context.TODO(), imageStreamName, metav1.DeleteOptions{})
+
+	// push a build to the image stream
+	buildConfig := &buildv1.Build{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: buildName,
+		},
+		Spec: buildv1.BuildSpec{
+			CommonSpec: buildv1.CommonSpec{
+				Source: buildv1.BuildSource{
+					Type: "Git",
+					Git: &buildv1.GitBuildSource{
+						URI: "https://github.com/cgwalters/fcos-derivation-example",
+						Ref: "rhcos",
+					},
+				},
+				Strategy: buildv1.BuildStrategy{
+					DockerStrategy: &buildv1.DockerBuildStrategy{},
+				},
+				Output: buildv1.BuildOutput{
+					To: &v1.ObjectReference{
+						Kind: "ImageStreamTag",
+						Name: imageStreamName + ":latest",
+					},
+				},
+			},
+		},
+	}
+	_, err = cs.BuildV1Interface.Builds(constants.MCONamespace).Create(context.TODO(), buildConfig, metav1.CreateOptions{})
+	require.Nil(t, err)
+	defer cs.BuildV1Interface.Builds(constants.MCONamespace).Delete(context.TODO(), buildName, metav1.DeleteOptions{})
+	helpers.WaitForBuild(t, cs, buildConfig.ObjectMeta.Name)
+
+	// pick a random worker node
+	unlabelFunc := helpers.LabelRandomNodeFromPool(t, cs, "worker", "node-role.kubernetes.io/infra")
+	defer unlabelFunc()
+	infraNode := helpers.GetSingleNodeByRole(t, cs, "infra")
+
+	// Run commands on the node:
```

Comment:
Actually I just remembered, I was hoping to not use `~/.docker/config.json` for rpm-ostree by default.  See https://github.com/coreos/rpm-ostree/pull/3239#issuecomment-975780989

But let's maybe just add a note like:

```suggestion
	// Write `~/.docker/config.json`.  This code will need to change in the future.
	// Probably the ostree stack will gain `/etc/ostree/container-auth.json`, and the MCD will need to write that.
```